From b91d0c2602e0b8759ada6f873cfb7c6b86b6d885 Mon Sep 17 00:00:00 2001
From: Shuzhen Wang <shuzhenwang@google.com>
Date: Tue, 29 Oct 2024 09:58:55 -0700
Subject: [PATCH 164/436] Camera: Init mMirrorModeForProducers for all
 constructors

mMirrorModeForProducers needs to match mGbps. Initialize it
in all of the OutputConfiguration constructors.

Flag: com.android.internal.camera.flags.mirror_mode_shared_surfaces
Test: Camera native CTS
Bug: 376159964
Change-Id: Idf14479701ae4e7a4ba09caea8b90524b7f3e2ea
---
 camera/camera2/OutputConfiguration.cpp | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/camera/camera2/OutputConfiguration.cpp b/camera/camera2/OutputConfiguration.cpp
index 48a733886e..daa26564f4 100644
--- a/camera/camera2/OutputConfiguration.cpp
+++ b/camera/camera2/OutputConfiguration.cpp
@@ -102,6 +102,12 @@ int OutputConfiguration::getMirrorMode(sp<IGraphicBufferProducer> surface) const
         return mMirrorMode;
     }
 
+    if (mGbps.size() != mMirrorModeForProducers.size()) {
+        ALOGE("%s: mGbps size doesn't match mMirrorModeForProducers: %zu vs %zu",
+                __FUNCTION__, mGbps.size(), mMirrorModeForProducers.size());
+        return mMirrorMode;
+    }
+
     // Use per-producer mirror mode if available.
     for (size_t i = 0; i < mGbps.size(); i++) {
         if (mGbps[i] == surface) {
@@ -350,6 +356,7 @@ OutputConfiguration::OutputConfiguration(sp<IGraphicBufferProducer>& gbp, int ro
     mStreamUseCase = ANDROID_SCALER_AVAILABLE_STREAM_USE_CASES_DEFAULT;
     mTimestampBase = TIMESTAMP_BASE_DEFAULT;
     mMirrorMode = MIRROR_MODE_AUTO;
+    mMirrorModeForProducers.push_back(mMirrorMode);
     mUseReadoutTimestamp = false;
     mFormat = HAL_PIXEL_FORMAT_IMPLEMENTATION_DEFINED;
     mDataspace = 0;
@@ -367,9 +374,9 @@ OutputConfiguration::OutputConfiguration(
     mColorSpace(ANDROID_REQUEST_AVAILABLE_COLOR_SPACE_PROFILES_MAP_UNSPECIFIED),
     mStreamUseCase(ANDROID_SCALER_AVAILABLE_STREAM_USE_CASES_DEFAULT),
     mTimestampBase(TIMESTAMP_BASE_DEFAULT),
-    mMirrorMode(MIRROR_MODE_AUTO), mUseReadoutTimestamp(false),
-    mFormat(HAL_PIXEL_FORMAT_IMPLEMENTATION_DEFINED), mDataspace(0),
-    mUsage(0) { }
+    mMirrorMode(MIRROR_MODE_AUTO), mMirrorModeForProducers(gbps.size(), mMirrorMode),
+    mUseReadoutTimestamp(false), mFormat(HAL_PIXEL_FORMAT_IMPLEMENTATION_DEFINED),
+    mDataspace(0), mUsage(0) { }
 
 status_t OutputConfiguration::writeToParcel(android::Parcel* parcel) const {
 
-- 
2.43.0

